src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+
+    if (prev_index > 0)
+    {
+        // New middle state, update previous
+        prev_state.next = new_index;
+        ReturnErrorOnFailure(prev_state.Save(mStorage));
+    }
+    else
+    {
+        // New first state
+        states.first = new_index;
+    }
+    // Update main list
+    states.count = static_cast<uint16_t>(actual_index + 1);
+    return states.Save(mStorage);
+
+    return CHIP_ERROR_INTERNAL;
+}
+
+CHIP_ERROR GroupDataProviderImpl::GetGroupState(size_t state_index, GroupState & out_state)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+
+    StateListData states;
+    VerifyOrReturnError(CHIP_NO_ERROR == states.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    // Fabric info found
+    // VerifyOrReturnError(state_index <= fabric.state_count, CHIP_ERROR_INVALID_ARGUMENT);
+
+    StateData state(states.first);
+    size_t actual_index = 0;
+
+    // Loop until the desired index
+    while (actual_index < states.count)
+    {
+        if (CHIP_NO_ERROR != state.Load(mStorage))
+        {
+            break;
+        }
+        if (actual_index == state_index)
+        {
+            // Target index found
+            out_state.fabric_index = state.fabric_index;
+            out_state.group        = state.group;
+            out_state.keyset_index = state.keyset_index;
+            return CHIP_NO_ERROR;
+        }
+        state.index = state.next;
+        actual_index++;
+    }
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR GroupDataProviderImpl::RemoveGroupState(size_t state_index)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+
+    StateListData states;
+
+    VerifyOrReturnError(CHIP_NO_ERROR == states.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    StateData state(states.first);
+    StateData prev_state;
+    size_t actual_index = 0;
+    size_t new_index    = 1;
+    size_t prev_index   = 0;
+    bool found          = false;
+
+    // Loop until the desired index
+    while (actual_index < states.count)
+    {
+        if (CHIP_NO_ERROR != state.Load(mStorage))
+        {
+            break;
+        }
+        if (new_index == state.index)
+        {
+            // Used index, keep looking
+            new_index++;
+        }
+        if (actual_index == state_index)
+        {
+            // Target index found
+            found = true;
+            break;
+        }
+        prev_index  = state.index;
+        prev_state  = state;
+        state.index = state.next;
+        actual_index++;
+    }
+
+    VerifyOrReturnError(found, CHIP_ERROR_KEY_NOT_FOUND);
+
+    if (states.count > 0)
+    {
+        states.count--;
+    }
+
+    ReturnErrorOnFailure(state.Delete(mStorage));
+    if (0 == prev_index)
+    {
+        // Remove first state
+        states.first = state.next;
+    }
+    else
+    {
+        // Remove intermediate state
+        prev_state.next = state.next;
+        ReturnErrorOnFailure(prev_state.Save(mStorage));
```

Comment:
If this fails, the entire storage is out of sync with orphan entries.

Reply:
There is no easy solution here. I have added an error check to the Delete(), which is performed first , but without a proper transaction, the recovery action itself may also fail, so we cannot ensure proper recovery.